clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1119,8 +1125,11 @@ public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration
         acquireAndEnsureOpen();
         long commitStart = time.nanoseconds();
         try {
-            CompletableFuture<Void> commitFuture = commit(offsets, true);
-            ConsumerUtils.getResult(commitFuture, time.timer(timeout));
+            Timer requestTimer = time.timer(timeout.toMillis());
+            // Commit with a timer to control how long the request should be retried until it
+            // gets a successful response or non-retriable error.
+            CompletableFuture<Void> commitFuture = commit(offsets, true, Optional.of(timeout.toMillis()));
+            ConsumerUtils.getResult(commitFuture, requestTimer);
```

Comment:
Let me see if I understand this correctly. It seems to me that we have two timers on this path. A first one here which bound the time waiting on the result. A second one in the background thread to no retry forever. Is my understanding correct?

If my understanding is correct, there is a chance that the first timer may timeout before the background thread had a change to complete the future. In this case, could we lose the real background exception?

Reply:
You got it right regarding the timers. And agree that if the case is that the request fails, is retried, and the retry response does not arrive in time, the request fails with a `TimeoutException` (there is even a [test](https://github.com/apache/kafka/blob/97eecdbac186a7b65228f7de665ebe552bb4a377/clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java#L471) to check that timeout is thrown in that case, seeing it from the angle that the response did not arrive in time), but agree that it would make more sense to propagate the original retriable exception. 

 That could definitely be improved. I already ensure [here](https://github.com/apache/kafka/blob/97eecdbac186a7b65228f7de665ebe552bb4a377/clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java#L564) that the original exception is propagated when expiring a request before retrying it (when a failed response is received), but this improvement would ensure that also happens in the case where a response is not received.